### DISCUSSION TITLE
Shell completions are too slow

### DISCUSSION THREAD

**Discussion by tuurep** [papis/papis#898] [Q&A] [answered]
Hey,

I'm trying to setup shell completions for papis with the help of this doc:

https://papis.readthedocs.io/en/latest/shell_completion.html

When I source what's in `<myProjectRoot>/scripts/shell_completion/click/papis.bash`, the tab completions now work e.g.

```
$ papis e<Tab>
edit     exec     explore  export
```

However it's quite painfully slow compared to readline/git equivalents (which I perceive as pretty much instant)

I'd say it's about 1 second.

So I'm wondering: is this a setup mistake, or do we have a chance to make the performance better?

It's probably related to [click](https://click.palletsprojects.com/en/8.1.x/shell-completion/), isn't it?

Also: really wish I could complete e.g. library names like:

```
$ papis -l <Tab>
long-library-name-foo    another-long-hehe    systematic-literature-reviews
```

So to speak, I'm interested in improving it in general.

**Comment by alexfikl** [comment, marked as answer]
I can reproduce this as well.. the shell completions on zsh are definitely annoyingly slow sometimes :(

The code for the completions is very simple (at least on zsh):
```sh
_papis_completion() {
    local -a completions
    local -a completions_with_descriptions
    local -a response
    (( ! $+commands[papis] )) && return 1

    response=("${(@f)$(env COMP_WORDS="${words[*]}" COMP_CWORD=$((CURRENT-1)) _PAPIS_COMPLETE=zsh_complete papis)}")

    for type key descr in ${response}; do
        if [[ "$type" == "plain" ]]; then
            if [[ "$descr" == "_" ]]; then
                completions+=("$key")
            else
                completions_with_descriptions+=("$key":"$descr")
            fi
        elif [[ "$type" == "dir" ]]; then
            _path_files -/
        elif [[ "$type" == "file" ]]; then
            _path_files -f
        fi
    done

    if [ -n "$completions_with_descriptions" ]; then
        _describe -V unsorted completions_with_descriptions -U
    fi

    if [ -n "$completions" ]; then
        compadd -U -V unsorted -a completions
    fi
}
```

As far as I can tell, this is slow for two reasons:
1. We call `_PAPIS_COMPLETE=zsh_complete papis` on every \<TAB\> to get the completions in real time from the command itself. This is obviously slower than hardcoding the options in some way.
2. Just calling `papis` to e.g. display the help is already quite slow. Last time I checked, this was mostly slow because we dynamically load commands (either from scripts in `~/.config/papis/scripts` or from entrypoints).

I'm not quite sure how to make this noticeably faster though. It might need some more serious profiling and maybe moving away from the click generated completions.

> Also: really wish I could complete e.g. library names like:

This would be very cool! I'm also not sure how to do that nicely. I assume this has to be added as a `click.Choice` to complete automagically, but that might create some dependency cycle in the code..

**Reply by alexfikl** [reply]
This may also do the trick for completing library names:
https://click.palletsprojects.com/en/latest/shell-completion/#custom-type-completion

**Reply by tuurep** [reply]
Thanks, very helpful to understand this! I've also since learned more about how shell completions actually work, was quite clueless when I first posted this :smile:

Yeah I'm wondering since papis is modeled after/inspired by the git style of command + subcommands, what I find overwhelmingly most useful to complete in git are long branch names, and similarly here would be long library names

Another idea I really wish papis had is aliasing subcommands in the same way git allows, for example in `~/.gitconfig`:

```gitconfig
[alias]
        s = status
	a = add
	c = commit
        ca = commit --amend
        p = push
        d = diff
        ds = diff --staged
        rs = restore --staged
        r = remote
        b = branch
        co = checkout
        l = log
```

Has this ever been brought up before, and do you like it? :) also I don't wanna be all over the place so let me know if this could be an issue or something.

**Reply by alexfikl** [reply]
> Yeah I'm wondering since papis is modeled after/inspired by the git style of command + subcommands, what I find overwhelmingly most useful to complete in git are long branch names, and similarly here would be long library names

I only have like 3-4 short libraries, so not very bothered by this, but I agree that it would be nice to complete those.

> Another idea I really wish papis had is aliasing subcommands in the same way git allows, for example in `~/.gitconfig`:

There is some support for some automatic shortcut aliases, e.g. `papis op` should work as `papis open`. The way that works is that it tries the suffix you typed in the known commands.. and if there's only one match it uses that.

It would be very cool to support something like what you suggest that's used in git. However, git is very flexible when it comes to aliases, so I'm not sure how much of that we want to implement.

> Has this ever been brought up before, and do you like it? :) also I don't wanna be all over the place so let me know if this could be an issue or something.

I don't recall seeing these suggestions before. Feel free to open some issues for them and we can mark it as an `enhancement` and see who's interested :grin:

If you want to give them a try, I'm happy to point you at the right chunk of code and help however I can!

**Reply by tuurep** [reply]
> e.g. `papis op` should work as `papis open`

Oh, so it does. I never even thought to try :smile: That's actually very nice, I use this heavily in `nmcli` and will be using here too.

Ok I can open issues, but sorry, what others are you thinking of than the gitconfig-style alias? The completions as discussed, but moved into an issue perhaps?

> If you want to give them a try, I'm happy to point you at the right chunk of code and help however I can!

I'm interested but also have a lot of different things to do so get easily distracted lately :) but certainly, please point me to the relevant code and there's a chance I could get going by reading it

E: also, didn't realize that the git aliasing is flexible or complex so I'll look more closely at that

**Reply by alexfikl** [reply]
> Ok I can open issues, but sorry, what others are you thinking of than the gitconfig-style alias? The completions as discussed, but moved into an issue perhaps?

Yes, I meant for the git-style aliases, the library completion and the slow completion in general. I don't think they're very related, so better add them as separate issues.

> I'm interested but also have a lot of different things to do so get easily distracted lately :) but certainly, please point me to the relevant code and there's a chance I could get going by reading it

Haha, can very much relate to that, so no worries! 

For the places where these things could be implemented:

1. Library completion (should be the easiest) should probably go here:
https://github.com/papis/papis/blob/95e0b5456afef60ee4f54c4aaec624244093c1e0/papis/commands/default.py#L139-L142
Using the click custom type completion framework. It could work by just making that a `click.Choice(papis.config.get_libs())`, but should be cleaner to use the recommended way.

2. For the git-style aliases it should probably go in the `get_command` method of the `ScriptLoaderGroup` here:
https://github.com/papis/papis/blob/95e0b5456afef60ee4f54c4aaec624244093c1e0/papis/commands/default.py#L66-L78
You can see there that we're trying to load entrypoints (that have `script.plugin`) and from various external scripts (either in your path or in `~/.config/papis/scripts`). You could probably extend that piece of code to also load the config section `[alias]` and try to run those instead.

It might be a bit of work to convince click to also handle `--flag` options correctly and anything else you might want. click is quite heavily used, so you might have some luck finding another project that does what you want :grin: 
 
> E: also, didn't realize that the git aliasing is flexible or complex so I'll look more closely at that

Yeah, it can get very fancy, but it should mostly be a search-and-replace sort of thing
https://www.eficode.com/blog/10-levels-of-git-aliases-beginner-to-intermediate-concepts
However, it also accepts external commands e.g. to run `gitk` on `git visual`:
```
[alias]
	visual = !gitk
```

**Comment by tuurep** [comment]
Thanks a lot. I'm opening issues but looks like I can't add the `enhancement` label myself so please add :)

**Reply by alexfikl** [reply]
Awesome! Thank you! I've added some labels there :grin:
